File: project/module-script/src/main/java/cn/fd/ratziel/module/script/api/ScriptExecutor.java
package cn.fd.ratziel.module.script.api;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.script.ScriptException;

/**
 * ScriptExecutor - 脚本执行者
 *
 * @author TheFloodDragon
 * @since 2024/7/14 21:26
 */
public interface ScriptExecutor {

    /**
     * 构建脚本
     *
     * @param script 脚本文本
     */
    @NotNull
    ScriptContent build(@NotNull String script);

    /**
     * 评估脚本
     *
     * @param script      原始脚本
     * @param environment 脚本运行环境
     * @throws ScriptException 当脚本评估中产生错误时抛出
     */
    @Nullable
    Object evaluate(@NotNull ScriptContent script, @NotNull ScriptEnvironment environment) throws ScriptException;

    /**
     * 评估脚本
     *
     * @param script      原始脚本
     * @param environment 脚本运行环境
     * @throws ScriptException 当脚本评估中产生错误时抛出
     */
    @Nullable
    default Object evaluate(@NotNull String script, @NotNull ScriptEnvironment environment) throws ScriptException {
        return evaluate(new LiteralScriptContent(script, this), environment);
    }

}

File: project/module-script/src/main/kotlin/cn/fd/ratziel/module/script/api/ScriptExecutor.kt
package cn.fd.ratziel.module.script.api

import javax.script.ScriptException

/**
 * ScriptExecutor - 脚本执行器
 *
 * @author TheFloodDragon
 * @since 2024/7/14 21:26
 */
interface ScriptExecutor {

    /**
     * 创建一个新的脚本编译器
     */
    fun compiler(): ScriptCompiler

    /**
     * 创建一个新的脚本评估器
     */
    fun evaluator(): ScriptEvaluator

    /**
     * 构建脚本
     *
     * @param  source 脚本源
     * @param environment 脚本环境
     */
    fun build(source: ScriptSource, environment: ScriptEnvironment = ScriptEnvironment()): Result<ScriptContent>

    /**
     * 评估脚本
     *
     * @param script      原始脚本
     * @param environment 脚本运行环境
     * @throws ScriptException 当脚本评估中产生错误时抛出
     */
    fun eval(script: ScriptContent, environment: ScriptEnvironment = ScriptEnvironment()): Result<Any?>

}
